docs(repo): update rulesets docs

heitortsergent opened the pull request:
Fixes #2231.

Preview link:

- https://docs.stoplight.io/docs/spectral/branches/docs%2Freorganize-rulesets/01baf06bdd05a-create-a-ruleset (former Rulesets page)
- https://docs.stoplight.io/docs/spectral/branches/docs%2Freorganize-rulesets/e5b9616d6d50c-rulesets (former Custom Rulesets page)

**Checklist**

- [ ] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

heitortsergent committed:
docs(repo): update intro to rulesets

heitortsergent committed:
docs(repo): expand on ruleset introduction

heitortsergent committed:
docs(repo): small update to intro

heitortsergent committed:
docs(repo): add section for ruleset properties

heitortsergent committed:
docs(repo): remove JSONPath Plus section

heitortsergent committed:
docs(repo): add create a ruleset section to rulesets

heitortsergent committed:
docs(repo): remove core/custom functions section

heitortsergent committed:
docs(repo): split up custom rulesets page

heitortsergent committed:
docs(repo): move formats to custom rulesets

heitortsergent committed:
docs(repo): move ruleset properties to custom rulesets

heitortsergent committed:
docs(repo): move extending rulesets section

heitortsergent committed:
docs(repo): move core rulesets section

heitortsergent committed:
docs(repo): small updates to custom rulesets overview

heitortsergent commented:
@philsturgeon would you mind taking a look at this and telling me what you think? I took a pass at reorganizing the Rulesets and Custom Rulesets page.

I think there's still some work to do here in the Custom Rulesets Overview page (some reorganizing and rewriting), but want to get your thoughts before I make more changes.

This link should work to see it in the docs view: https://docs.stoplight.io/docs/spectral/branches/docs%2Freorganize-rulesets/01baf06bdd05a-rulesets

philsturgeon commented:
Custom Rulesets starts with this which no longer makes sense in the move: 

> Customizing existing rulesets might be all you need at first, but at some point, you will want to make a custom ruleset.

Let's start from scratch.

Lets look through the various keywrods that make up a ruleset, so you can learn to tweak a distributed ruleset to work for you, or make your own ruleset from scratch to power your organizations API Style Guide or some other custom situation.

philsturgeon commented:
Generally this is looking amazing! Using a group is a fantastic way to break it down.

philsturgeon commented:
Btw I noticed an old mistake I made, I had `given: $.paths.[*][get,post,put,delete,options]` but it should be `given: $.paths[*][get,post,put,delete,options]`.

From @P0lip: 

> `$.paths.[*]` acts as `$.paths..[*]` so it goes deep - it's incorrect if you only want to target an operation object.

heitortsergent committed:
docs(repo): update resolved rule example

heitortsergent commented:
@philsturgeon thanks!! I fixed the `given` example you mentioned.

I'll go ahead and update the intro to Custom Rulesets and make some more changes then. I'll tag you again once it's ready for another look. :)

heitortsergent committed:
docs(repo): add lint instruction to rulesets

heitortsergent committed:
docs(repo): add link to other rulesets

heitortsergent committed:
docs(repo): reorganize custom rulesets, update intro

heitortsergent committed:
docs(repo): small clarification for formats

heitortsergent committed:
docs(repo): remove link

heitortsergent committed:
docs(repo): remove unnecessary words

heitortsergent committed:
docs(repo): fix broken link

heitortsergent committed:
docs(repo): remove core rulesets section

heitortsergent committed:
docs(repo): update rules section

heitortsergent committed:
docs(repo): delete rules header and move content

heitortsergent committed:
docs(repo): add properties to ruleset and rules list

heitortsergent committed:
docs(repo): change Rulesets title to Create a Ruleset

heitortsergent committed:
docs(repo): update headings

heitortsergent committed:
docs(repo): update toc Custom Rulesets to Rulesets

heitortsergent committed:
docs(repo): break up rules into its own page

heitortsergent committed:
docs(repo): cleanup md

heitortsergent committed:
docs(repo): update ruleset properties section

heitortsergent committed:
docs(repo): update rules properties

heitortsergent committed:
docs(repo): rewording

heitortsergent commented:
@philsturgeon I think it's ready for another look if you have some time? :D 

- Renamed the Rulesets page to Create a Ruleset - I think that makes more sense in terms of that page being a part of the "Getting Started" section
- Renamed "Custom Rulesets" to "Rulesets"
- Updated the intro to "Rulesets" page to emphasize "Custom Ruleset" a bit less (like you mentioned in your previous comment)
- Split up "Rules" into its own separate page - I was struggling a little bit with having the Ruleset Properties and Rule Properties in the same page, I think this might work better

I don't love having the Rulesets group now with: Overview, Rules, and then specific pages for certain Ruleset Properties (Aliases, etc.), but I'm not sure having a subgroup in there is a good idea... I'm thinking if this seems like a better organization, I could also stop at this point for this PR and just fix the links, and we can play around with a bit more in a separate PR.

Let me know what you think! 🙏

heitortsergent committed:
docs(repo): rename files

heitortsergent committed:
docs(repo): update toc

heitortsergent committed:
docs(repo): fix link

pamgoodrich commented on docs/getting-started/3-rulesets.md:
```diff
@@ -1,161 +1,60 @@
+- `spectral:asyncapi` - [AsyncAPI v2 rules](./5-asyncapi.md)
 
-Extends can reference any [distributed ruleset](../guides/7-sharing-rulesets.md). It can be a single string, or an array of strings, and can contain either local file paths, URLs, or even npm modules.
+To create a ruleset that extends both rulesets, in your terminal run:
```
Punctuation is a little off here. I think you need a comma between "in your terminal" and "run"

pamgoodrich commented on docs/getting-started/3-rulesets.md:
```diff
@@ -1,161 +1,60 @@
-- `json-schema-2020-12` (`$schema` says this is JSON Schema 2020-12)
-
-Specifying the format is optional, so you can completely ignore this if all the rules you are writing apply to any document you lint, or if you have specific rulesets for different formats. If you'd like to use one ruleset for multiple formats, the `formats` key is here to help.
+The newly created ruleset file can then be used to lint any OpenAPI v2/v3 or AsyncAPI descriptions by using the `spectral lint` command:
```
descriptions by using > descriptions using

pamgoodrich commented on docs/getting-started/3-rulesets.md:
```diff
@@ -1,161 +1,60 @@
+Breaking down each part of the rule:
 
-## Core Rulesets
+- `description` and `message` will help users quickly understand what the goal of the rule is
```
Use active voice here. 

"will help" > "help"  in two instances

pamgoodrich commented on docs/guides/4-custom-rulesets.md:
```diff
@@ -1,247 +1,102 @@
-Add your own rules under the `rules` property in your `.spectral.yml`, or another ruleset file.
+- `rules` (required): An array of rules. See [Rules](./4a-rules.md) for more details.
+- `extends` (optional): A reference to other rulesets. Used to extend and customize existing rulesets. See [Extends](./4b-extends.md) for more details.
+- `formats` (optional): The format that the ruleset should apply to. For example `oas3` for any OpenAPI v3.x descriptions. Can be applied at the ruleset and/or rule level. See [Formats](#formats) for more details.
```
For example `oas3` > For example, `oas3`

pamgoodrich commented on docs/guides/4-custom-rulesets.md:
```diff
@@ -1,247 +1,102 @@
+- `formats` (optional): The format that the ruleset should apply to. For example `oas3` for any OpenAPI v3.x descriptions. Can be applied at the ruleset and/or rule level. See [Formats](#formats) for more details.
+- `documentationUrl` (optional): A URL that contains more information about the ruleset and rules in it. Can help provide users more context on why the ruleset exists and how it should be used. See [Documentation URL](#documentation-url) for more details.
+- `parserOptions` (optional): Can be used to tune the severity of duplicate keys or invalid values in your ruleset. See [Parsing Options](#parsing-options) for more details.
+- `aliases` (optional): An array of key-value pairs that can be used to define commonly used JSONPath expressions, to be reused across a ruleset. See [Aliases](./4c-aliases.md) for more details.
```
Here, I think you can remove the comma between "expressions" and "to be reused"

pamgoodrich commented on docs/guides/4-custom-rulesets.md:
```diff
@@ -1,247 +1,102 @@
-If your rule needs to access the raw `$ref` reference values, you can set `resolved: false` to allow the rule to receive the raw un-resolved version of the document. Otherwise `resolved: true` is the default.
-
-Here's an example of a rule that can access `$ref` values:
+Specifying the format is optional, so you can completely ignore this if all the rules you are writing apply to any document you lint, or if you have specific rulesets for different formats. If you'd like to use one ruleset for multiple formats, the `formats` key is here to help.
```
Maybe change "the formats key is here to help" to "use the formats key."?

heitortsergent replied:
There was also some repetition here so I cleaned it up a bit.

pamgoodrich commented on docs/guides/4-custom-rulesets.md:
````diff
@@ -281,361 +136,17 @@ rules:
-    User:
-      $ref: "./User.yaml"
-```
+If you do not care about duplicate keys or invalid values (such as non-string mapping keys in YAML), you can tune their severity using the `parserOptions` setting.
````
do not > don't :)

pamgoodrich commented on docs/guides/4a-rules.md:
```diff
@@ -0,0 +1,165 @@
+# Rules
+
+Rules are at the core of how rulesets work, so let's look at how to create a rule and its properties.
```
Rules are at the core of how rulesets work > Rules are the core of results.

pamgoodrich commented on docs/guides/4a-rules.md:
```diff
@@ -0,0 +1,165 @@
+- `given` (required): The part of the document the rule should be applied to. Uses the [JSONPath](https://goessner.net/articles/JsonPath/index.html) syntax.
+- `then` (required): Describes which function should be applied to the `given` part of the document. Can be used with a [core function](../reference/functions.md) or [custom function](./5-custom-functions.md).
+- `description` (optional): A short description of the rule.
+- `message` (optional): A message that will be displayed in the `spectral lint` output. Can be customized to use placeholder values that are evaluated at runtime, such as `{{description}}` or `{{error}}`.
```
that will be > that is

heitortsergent replied:
Changed to "that's" :D

pamgoodrich commented on docs/guides/4a-rules.md:
```diff
@@ -0,0 +1,165 @@
+
+### Given
+
+The `given` property is conceptually quite like a selector in CSS, in that it picks the part of the document to apply rules to.
```
I'm not a fan of the word "picks", but it's writer's choice. I'd probably use "indicates"

heitortsergent replied:
I like "indicates" better 👍

pamgoodrich commented on docs/guides/4a-rules.md:
```diff
@@ -0,0 +1,165 @@
+
+The `given` property is conceptually quite like a selector in CSS, in that it picks the part of the document to apply rules to.
+
+It has a specific syntax known as [JSONPath](https://goessner.net/articles/JsonPath/index.html), which if you are familiar with XPath is quite similar. JSONPath is not yet a standard (it [will be](https://tools.ietf.org/html/draft-normington-jsonpath-00) someday), and has a few competing implementations. Spectral uses [nimma](https://www.npmjs.com/package/nimma) as its main implementation, and so ...
```
"ensure a good" > "ensure good"

pamgoodrich commented on docs/guides/4a-rules.md:
````diff
@@ -0,0 +1,165 @@
+  function: truthy
+```
+
+The `field` keyword is optional and is used to apply the function to a specific property in an object. If omitted the function will be applied to the entire target of the `given` JSONPath. The value can also be `@key` to apply the rule to the keys of an object.
````
If omitted, the

pamgoodrich commented on docs/guides/4a-rules.md:
```diff
@@ -0,0 +1,165 @@
+
+### Message
+
+To help you create meaningful messages for results, Spectral comes with a couple of placeholders that are evaluated at runtime.
```
I'd remove "a couple of" since that typically means two and we have five.

pamgoodrich commented on docs/guides/4a-rules.md:
```diff
@@ -0,0 +1,165 @@
+to). While this is typically the desired behavior, there are some use cases
+where you may need to run a rule on the "raw" un-resolved document.
+
+For example, if you want to enforce conventions on the folder structure used for
```
This sentence is incomplete. You could say: "For example, use `resolved` to enforce..."

pamgoodrich commented on docs/guides/4b-extends.md:
````diff
@@ -0,0 +1,49 @@
+  - some-npm-module # note that this would be treated as any other npm package, therefore it has to be placed under node_modules and have a valid package.json.
+```
+
+The `extends` keyword can be combined with extra rules in order to extend and override rulesets. Learn more about that in [custom rulesets](../guides/4-custom-rulesets.md).
````
in order to > to

pamgoodrich commented on docs/guides/4b-extends.md:
```diff
@@ -0,0 +1,49 @@
+
+This provides a new description, but anything can be changed.
+
+If you're just looking to change the severity of the rule, there is a handy shortcut.
```
there is > there's

pamgoodrich commented on docs/guides/4b-extends.md:
````diff
@@ -0,0 +1,49 @@
+  operation-success-response: warn
+```
+
+Available severity levels are `error`, `warn`, `info`, `hint`, and `off`.
````
This is the first time we mention off as a severity level. For style guides, that's called disabled. Do we need to add it to the other places where we mention severity?

heitortsergent replied:
Good catch. I replaced this so instead of having this information duplicated it links to the "Severity" topic in 4a.rules.

pamgoodrich commented on docs/guides/4c-aliases.md:
```diff
@@ -0,0 +1,116 @@
+## Aliases
+
+Targeting certain parts of an OpenAPI spec is powerful but it can become cumbersome to write and repeat complex JSONPath expressions across various rules.
+Define aliases for commonly used JSONPath expressions on a global level which can then be reused across the ruleset.
```
Define aliases for commonly used JSONPath expressions on a global level which can then be reused across the ruleset.
>
Define aliases for commonly used JSONPath expressions on a global level and then reuse them across the ruleset.

pamgoodrich commented on docs/guides/4c-aliases.md:
````diff
@@ -0,0 +1,116 @@
+        then:
+          function: falsy
+```
+
````
Not sure what this means?

pamgoodrich commented on docs/guides/4d-overrides.md:
```diff
@@ -0,0 +1,125 @@
+## Overrides
+
+Previously Spectral supported exceptions, which were limited in their ability to target particular rules on specific files or parts of files, or change parts of a rule. Overrides is the much more powerful version of exceptions, with the ability to customize ruleset usage for different files and projects without having to duplicate any rules.
```
Overrides is > Overrides are

pamgoodrich commented on docs/guides/4d-overrides.md:
```diff
@@ -0,0 +1,125 @@
+
+### Caveats
+
+Please bear in mind that overrides are only applied to the _root_ documents. If your documents have any external dependencies, i.e. $refs, the overrides won't apply.
```
Remove Please and maybe just say, "Overrides are only applied..." The heading already indicates that this is a caveat.

pamgoodrich commented on docs/guides/4e-recommended.md:
````diff
@@ -0,0 +1,41 @@
+extends: [[spectral:oas, all]]
+```
+
+You can do this with your rulesets, and slide new rules in as not recommended for a while so that only the most interested active API designers/developers get them at first, then eventually roll them out to everyone if they are well received.
````
You can do this with your rulesets, and initially set new rules as "not recommended" so that the most interested active API designers/developers use them at first. Then, eventually roll them out to everyone if they are well received.

pamgoodrich commented on docs/guides/4e-recommended.md:
````diff
@@ -0,0 +1,41 @@
+  operation-operationId-unique: off
+```
+
+The example above will run all of the rules defined in the `spectral:oas` ruleset (rather than the default behavior that runs only the recommended ones), with one exception - we turned `operation-operationId-unique` off.
````
will run > runs

pamgoodrich replied:
exception - we turned `operation-operationId-unique` off. > exception; 
 `operation-operationId-unique` is disabled.

pamgoodrich replied:
Or "set to off" instead of "disabled"

pamgoodrich commented on docs/guides/4e-recommended.md:
```diff
@@ -0,0 +1,41 @@
+
+### Enabling Rules
+
+Sometimes you might want to apply a limited number of rules from another ruleset. To do this, use the `extends` property with `off` as the second argument. This will avoid running any rules from the extended ruleset as they will all be disabled. Then you can pick and choose which rules you would like to enable.
```
This will avoid > This avoids

pamgoodrich commented on docs/guides/4e-recommended.md:
````diff
@@ -0,0 +1,41 @@
+  operation-operationId-unique: true
+```
+
+The example above will only run the rule `operation-operationId-unique` that we enabled since we passed `off` to disable all rules by default when extending the `spectral:oas` ruleset.
````
"above will only run" > "above only runs"

pamgoodrich commented on docs/guides/4f-js-rulesets.md:
```diff
@@ -0,0 +1,110 @@
+## JavaScript Ruleset Format
+
+Spectral v6.0 added support for a JavaScript ruleset format, similar to the JSON and YAML formats.
```
I think I asked before. Do we need the v6.0 here?

heitortsergent replied:
You did, and I tagged other folks in that comment but didn't get a reply, so it stayed in there since I wasn't sure if it was ok to remove.

pamgoodrich commented on docs/guides/4f-js-rulesets.md:
```diff
@@ -0,0 +1,110 @@
+
+Spectral v6.0 added support for a JavaScript ruleset format, similar to the JSON and YAML formats.
+
+This has a few benefits: it lets you explicitly load formats or rulesets to get control over versioning, you can load common functions from popular JS libraries, and in general feels a lot more welcoming to developers experienced with JavaScript, especially when it comes to working with custom functions.
```
This is a bit of a long sentence. Maybe break it into bullets? Writer's choice.

pamgoodrich commented on docs/guides/4f-js-rulesets.md:
````diff
@@ -0,0 +1,110 @@
+npm install --save @stoplight/spectral-formats
+```
+
+Installing these packages is not required for creating a JavaScript ruleset, but we'll use them in our example to create some common rules used with Spectral and to target a specific OpenAPI format.
````
we'll > you'll ?

our > the ?

pamgoodrich commented on docs/guides/4f-js-rulesets.md:
````diff
@@ -0,0 +1,110 @@
+touch spectral.js
+```
+
+And inside the file, let's create a couple rules:
````
I've seen the word "let's" a few times. Not a huge fan and it breaks our Vale rule. You could just remove it and say, "Inside the file, create a couple of rules:"

You do need the "of" between couple and rules, though.

pamgoodrich commented on docs/guides/4f-js-rulesets.md:
````diff
@@ -0,0 +1,110 @@
+};
+```
+
+For those of you using custom functions, the keywords `functions` & `functionOptions` have been removed, as they were designed to help Spectral find your functions. Now functions are passed as a variable, instead of using a string that contains the name like the JSON/YAML formats.
````
For those of you using... > If you use custom functions

pamgoodrich replied:
& > and

pamgoodrich reviewed (dismissed):
Great updates!

heitortsergent committed:
docs(repo): pam feedback

heitortsergent committed:
Merge branch 'develop' into docs/reorganize-rulesets

heitortsergent committed:
Merge branch 'develop' into docs/reorganize-rulesets

heitortsergent committed:
Merge branch 'develop' into docs/reorganize-rulesets
